File: assignments/FrequentPatterns/src/main/java/edu/yu/bioinfo/PatternCount.java
//PatternCount.java
package edu.yu.bioinfo;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class PatternCount {

    /**
     * Counts the occurrences of a pattern in a given text.
     *
     * @param text    The input text.
     * @param pattern The pattern to count.
     * @return The count of occurrences of the pattern in the text.
     * @throws IllegalArgumentException If the pattern is empty or longer than the
     *                                  text.
     */
    public static int patternCount(String text, String pattern) {
        // Validate input
        if (pattern.isEmpty()) {
            return 0; // If the pattern is empty, it cannot have occurrences
        }

        // Check if Pattern is longer than text, if it is that is an illegal argument
        if (pattern.length() > text.length()) {
            throw new IllegalArgumentException("Pattern cannot be longer than Text");
        }
        
        int count = 0;
        // Iterate through the text to find occurrences of the pattern
        for (int i = 0; i <= text.length() - pattern.length(); i++) {
            // Check if the substring of the text matches the pattern
            if (text.substring(i, i + pattern.length()).equals(pattern)) {
                count++;
            }
        }
        return count;
    }
    /**
     * Reads input from a text file, runs patternCount, and prints the result.
     *
     * @param args Command-line arguments (not used in this example).
     */
    public static void main(String[] args) {
        // Read input from a text file
        String fileName = "data/Vibrio_cholerae.txt"; // Provide the actual file name
        String text = "";
        String pattern = "TGATCA";

        try {
            Scanner scanner = new Scanner(new File(fileName));

            // Assuming the first line is the text and the second line is the pattern
            if (scanner.hasNextLine()) {
                text = scanner.nextLine();
            }
            //if (scanner.hasNextLine()) {
                //pattern = scanner.nextLine();
            //}

            scanner.close();
        } catch (FileNotFoundException e) {
            // Handle file not found exception
            System.err.println("File not found: " + e.getMessage());
            return;
        }

        // Output the result of patternCount
        int count = patternCount(text, pattern);
        System.out.println("Occurrences of pattern '" + pattern + "' in the text: " + count);
    }
}

File: assignments/FrequentPatterns/src/main/python/edu/yu/bioinfo/PatternCount.py
# PatternCount.py

def PatternCount(Text, Pattern):
    """
    Counts the occurrences of a pattern in a given text.

    Args:
    - Text (str): The input text.
    - Pattern (str): The pattern to count.

    Returns:
    - int: The count of occurrences of the pattern in the text.

     Raises:
    - ValueError: If Text or Pattern is empty, or if Pattern is longer than Text.
    - TypeError: If Text or Pattern is not a string.
    """

   # Validate input strings
    if not isinstance(Text, str) or not isinstance(Pattern, str):
        raise TypeError("Text and Pattern must be strings")

    # Check if Text and Pattern are non-empty strings
    if not Text or not Pattern:
        raise ValueError("Text and Pattern cannot be empty")
    
    # Check if Pattern is not longer than Text
    if len(Pattern) > len(Text):
        raise ValueError("Pattern length cannot be greater than Text length")
    
    # Initialize the count of occurences
    count = 0

    try:
        # Iterate through each possible starting position of a substring of length Pattern in Text
        for i in range(len(Text) - len(Pattern) + 1):
            # Check if the substring from position i to i + len(Pattern) is equal to the given Pattern
            if Text[i:i+len(Pattern)] == Pattern:
                # If equal, increment the count
                count += 1
    except Exception as e:
        # Raise an exception with a descriptive error message
        raise ValueError(f"Error counting pattern occurrences: {e}")

    # Return the final count of occurrences
    return count

'''
This code defines a function PatternCount that counts the occurrences of a given pattern (Pattern) in a given text (Text). 
The function uses a sliding window approach to iterate through the text and check for matches with the pattern. 
If a match is found, the count is incremented. Any exceptions during the process are caught and an error message is printed. 
The final count is returned.
'''

